File: 543-diameter-of-binary-tree/543-diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int height(TreeNode *root, int &diameter){
        if(!root)return 0;
        
        int left_height = height(root->left, diameter);
        int right_height = height(root->right, diameter);
        
        diameter = max(diameter, left_height + right_height);
        return max(left_height,right_height)+1;
    }
    
    int diameterOfBinaryTree(TreeNode* root) {
        int diameter =0;
        height(root,diameter);
        return diameter;
    }
};

File: 543-diameter-of-binary-tree/543-diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    public int height(TreeNode root){
        if(root == null)return 0;
        
        int lh= height(root.left);
        int rh = height(root.right);
        return Math.max(lh, rh)  + 1;
    }
    
    public int diameterOfBinaryTree(TreeNode root) {
        if(root == null)return 0;
        
        int lh = height(root.left);
        int rh = height(root.right);
        
        int withRoot = lh + rh;
        int left = diameterOfBinaryTree(root.left);
        int right = diameterOfBinaryTree(root.right);
        
        return Math.max(withRoot, Math.max(left, right));
    }
}